### ISSUE TITLE
WIP

### ISSUE THREAD

**Issue by kenz-gelsoft** [kenz-gelsoft/codepositive#1]

**Comment by kenz-gelsoft** [comment]
```
~/src/vscode> ./scripts/code-server.sh --launch
~/src/vscode ~/src/vscode
yarn run v1.22.22
$ node build/lib/electron
node:internal/modules/cjs/loader:1148
  throw err;
  ^

Error: Cannot find module 'ternary-stream'
Require stack:
- /boot/home/src/vscode/build/lib/util.js
- /boot/home/src/vscode/build/lib/electron.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1145:15)
    at Module._load (node:internal/modules/cjs/loader:986:27)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/boot/home/src/vscode/build/lib/util.js:38:23)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/boot/home/src/vscode/build/lib/util.js',
    '/boot/home/src/vscode/build/lib/electron.js'
  ]
}

Node.js v20.15.1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
yarn run v1.22.22
$ node --max-old-space-size=8192 ./node_modules/gulp/bin/gulp.js node
Error: Cannot find module 'ternary-stream'
Require stack:
- /boot/home/src/vscode/build/lib/util.js
- /boot/home/src/vscode/build/gulpfile.js
- /boot/home/src/vscode/gulpfile.js
- /boot/home/src/vscode/node_modules/gulp-cli/lib/shared/require-or-import.js
- /boot/home/src/vscode/node_modules/gulp-cli/lib/versioned/^4.0.0/index.js
- /boot/home/src/vscode/node_modules/gulp-cli/index.js
- /boot/home/src/vscode/node_modules/gulp/bin/gulp.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1145:15)
    at Module._load (node:internal/modules/cjs/loader:986:27)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/boot/home/src/vscode/build/lib/util.js:38:23)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/boot/home/src/vscode/build/lib/util.js',
    '/boot/home/src/vscode/build/gulpfile.js',
    '/boot/home/src/vscode/gulpfile.js',
    '/boot/home/src/vscode/node_modules/gulp-cli/lib/shared/require-or-import.js',
    '/boot/home/src/vscode/node_modules/gulp-cli/lib/versioned/^4.0.0/index.js',
    '/boot/home/src/vscode/node_modules/gulp-cli/index.js',
    '/boot/home/src/vscode/node_modules/gulp/bin/gulp.js'
  ]
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
~/src/vscode
./scripts/code-server.sh: line 26: /boot/home/src/vscode/.build/node/v20.14.0/haiku-x64/node: No such file or directory
```

**Comment by kenz-gelsoft** [comment]
```
~/src/vscode> yarn
yarn install v1.22.22
$ node build/npm/preinstall.js
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning "@microsoft/1ds-core-js > @microsoft/applicationinsights-core-js@2.8.15" has unmet peer dependency "tslib@*".
warning " > tsec@0.2.7" has unmet peer dependency "@bazel/bazelisk@>=1.7.5".
warning " > tsec@0.2.7" has unmet peer dependency "@bazel/concatjs@>=5.3.0".
[4/4] Building fresh packages...
[1/21] ⡀ @parcel/watcher
[10/21] ⡀ kerberos
[11/21] ⡀ native-is-elevated
[6/21] ⡀ @vscode/sqlite3
error /boot/home/src/vscode/node_modules/kerberos: Command failed.
Exit code: 1
Command: prebuild-install --runtime napi || node-gyp rebuild
Arguments: 
Directory: /boot/home/src/vscode/node_modules/kerberos
Output:
prebuild-install warn This package does not support N-API version 30.1.2
gyp info it worked if it ends with ok
gyp info using node-gyp@10.1.0
gyp info using node@20.15.1 | haiku | x64
gyp info find Python using Python version 3.10.14 found at "/bin/python3"

gyp info spawn /bin/python3
gyp info spawn args [
gyp info spawn args '/boot/system/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args 'binding.gyp',
gyp info spawn args '-f',
gyp info spawn args 'make',
gyp info spawn args '-I',
gyp info spawn args '/boot/home/src/vscode/node_modules/kerberos/build/config.gypi',
gyp info spawn args '-I',
gyp info spawn args '/boot/system/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args '-I',
gyp info spawn args '/boot/home/config/cache/node-gyp/30.1.2/include/node/common.gypi',
gyp info spawn args '-Dlibrary=shared_library',
gyp info spawn args '-Dvisibility=default',
gyp info spawn args '-Dnode_root_dir=/boot/home/config/cache/node-gyp/30.1.2',
gyp info spawn args '-Dnode_gyp_dir=/boot/system/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args '-Dnode_lib_file=/boot/home/config/cache/node-gyp/30.1.2/<(target_arch)/node.lib',
gyp info spawn args '-Dmodule_root_dir=/boot/home/src/vscode/node_modules/kerberos',
gyp info spawn args '-Dnode_engine=v8',
gyp info spawn args '--depth=.',
gyp info spawn args '--no-parallel',
gyp info spawn args '--generator-output',
gyp info spawn args 'build',
gyp info spawn args '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/boot/home/src/vscode/node_modules/kerberos/build'
  CXX(target) Release/obj.target/kerberos/src/kerberos.o
In file included from ../src/kerberos_common.h:13,
                 from ../src/kerberos.h:13,
                 from ../src/kerberos.cc:1:
../src/win32/kerberos_sspi.h:22:10: fatal error: windows.h: No such file or directory
   22 | #include <windows.h>
      |          ^~~~~~~~~~~
compilation terminated.
make: *** [Release/obj.target/kerberos/src/kerberos.o] Error 1
kerberos.target.mk:123: recipe for target 'Release/obj.target/kerberos/src/kerberos.o' failed
make: Leaving directory '/boot/home/src/vscode/node_modules/kerberos/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack at ChildProcess.<anonymous> (/boot/system/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:209:23)
gyp ERR! System Haiku 1
gyp ERR! command "/boot/system/bin/node" "/boot/system/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /boot/home/src/vscode/node_modules/kerberos




Error: ENOENT: no such file or directory, open '/usr/local/share/.yarnrc'
```

**Comment by kenz-gelsoft** [comment]
https://esbuild.github.io/getting-started/#build-from-source

Maybe I should cross-build for `haiku x64 LE` platform.
```
error /boot/home/src/vscode/build/node_modules/esbuild: Command failed.
Exit code: 1
Command: node install.js
Arguments: 
Directory: /boot/home/src/vscode/build/node_modules/esbuild
Output:
/boot/home/src/vscode/build/node_modules/esbuild/install.js:77
    throw new Error(`Unsupported platform: ${platformKey}`);
          ^

Error: Unsupported platform: haiku x64 LE
    at pkgAndSubpathForCurrentPlatform (/boot/home/src/vscode/build/node_modules/esbuild/install.js:77:11)
    at checkAndPreparePackage (/boot/home/src/vscode/build/node_modules/esbuild/install.js:253:28)
    at Object.<anonymous> (/boot/home/src/vscode/build/node_modules/esbuild/install.js:280:1)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:174:12)
    at node:internal/main/run_main_module:28:49
```

**Comment by kenz-gelsoft** [comment]
https://github.com/microsoft/vscode/wiki/Cross-Compiling-for-Debian-Based-Linux

might help.

**Comment by kenz-gelsoft** [comment]
yarn; yarn compile on macOS, and transferred vscode source tree, I could launch code-server with following command line.

```
/名称未設定/codepositive> VSCODE_DEV=1 node out/server-main.js 
Server bound to 127.0.0.1:8000 (IPv4)
Extension host agent listening on 8000

[21:33:46] 




Error: Not an executable
    at Module._extensions..node (node:internal/modules/cjs/loader:1454:18)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at bindings (/名称未設定/codepositive/remote/node_modules/bindings/bindings.js:112:48)
    at Object.<anonymous> (/名称未設定/codepositive/remote/node_modules/@vscode/spdlog/index.js:3:35)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at nodeRequire (file:///Users/kenz/src/vscode/src/vs/loader.js:1025:12)
    at NodeScriptLoader.load (file:///Users/kenz/src/vscode/src/vs/loader.js:823:22)
    at OnlyOnceScriptLoader.load (file:///Users/kenz/src/vscode/src/vs/loader.js:581:23)
    at loadNextPath (file:///Users/kenz/src/vscode/src/vs/loader.js:1620:25)
    at Object.errorback (file:///Users/kenz/src/vscode/src/vs/loader.js:1628:7)
    at OnlyOnceScriptLoader.triggerErrorback (file:///Users/kenz/src/vscode/src/vs/loader.js:594:24)
    at file:///Users/kenz/src/vscode/src/vs/loader.js:581:107
    at NodeScriptLoader.load (file:///Users/kenz/src/vscode/src/vs/loader.js:826:6)
    at OnlyOnceScriptLoader.load (file:///Users/kenz/src/vscode/src/vs/loader.js:581:23)
    at loadNextPath (file:///Users/kenz/src/vscode/src/vs/loader.js:1620:25)
    at ModuleManager._loadModule (file:///Users/kenz/src/vscode/src/vs/loader.js:1632:4)
    at ModuleManager._resolve (file:///Users/kenz/src/vscode/src/vs/loader.js:1715:11)
    at ModuleManager.defineModule (file:///Users/kenz/src/vscode/src/vs/loader.js:1369:9)
    at ModuleManager._relativeRequire (file:///Users/kenz/src/vscode/src/vs/loader.js:1403:9)
    at result (file:///Users/kenz/src/vscode/src/vs/loader.js:1574:17)
    at /名称未設定/codepositive/out/vs/platform/log/node/spdlogLog.js:45:74
    at new Promise (<anonymous>)
    at createSpdLogLogger (/名称未設定/codepositive/out/vs/platform/log/node/spdlogLog.js:45:35)
    at SpdLogLogger._createSpdLogLogger (/名称未設定/codepositive/out/vs/platform/log/node/spdlogLog.js:120:34)
    at new SpdLogLogger (/名称未設定/codepositive/out/vs/platform/log/node/spdlogLog.js:110:48)
    at LoggerService.doCreateLogger (/名称未設定/codepositive/out/vs/platform/log/node/loggerService.js:11:20)
    at LoggerService.createLogger (/名称未設定/codepositive/out/vs/platform/log/common/log.js:408:31)
    at setupServerServices (/名称未設定/codepositive/out/vs/server/node/serverServices.js:51:38)
    at createServer (/名称未設定/codepositive/out/vs/server/node/remoteExtensionHostAgentServer.js:600:103)
    at async file:///Users/kenz/src/vscode/src/server-main.js:91:20
    at async Server.<anonymous> (file:///Users/kenz/src/vscode/src/server-main.js:179:3) {
  code: 'ERR_DLOPEN_FAILED',
  phase: 'loading',
  moduleId: '@vscode/spdlog',
  neededBy: [ '===anonymous2===' ]
}
[21:33:46] Error: Unsupported platform
    at Object.<anonymous> (/名称未設定/codepositive/remote/node_modules/@vscode/deviceid/dist/index.js:24:11)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at nodeRequire (file:///Users/kenz/src/vscode/src/vs/loader.js:1025:12)
    at NodeScriptLoader.load (file:///Users/kenz/src/vscode/src/vs/loader.js:823:22)
    at OnlyOnceScriptLoader.load (file:///Users/kenz/src/vscode/src/vs/loader.js:581:23)
    at loadNextPath (file:///Users/kenz/src/vscode/src/vs/loader.js:1620:25)
    at Object.errorback (file:///Users/kenz/src/vscode/src/vs/loader.js:1628:7)
    at OnlyOnceScriptLoader.triggerErrorback (file:///Users/kenz/src/vscode/src/vs/loader.js:594:24)
    at file:///Users/kenz/src/vscode/src/vs/loader.js:581:107
    at NodeScriptLoader.load (file:///Users/kenz/src/vscode/src/vs/loader.js:826:6)
    at OnlyOnceScriptLoader.load (file:///Users/kenz/src/vscode/src/vs/loader.js:581:23)
    at loadNextPath (file:///Users/kenz/src/vscode/src/vs/loader.js:1620:25)
    at ModuleManager._loadModule (file:///Users/kenz/src/vscode/src/vs/loader.js:1632:4)
    at ModuleManager._resolve (file:///Users/kenz/src/vscode/src/vs/loader.js:1715:11)
    at ModuleManager.defineModule (file:///Users/kenz/src/vscode/src/vs/loader.js:1369:9)
    at ModuleManager._relativeRequire (file:///Users/kenz/src/vscode/src/vs/loader.js:1403:9)
    at result (file:///Users/kenz/src/vscode/src/vs/loader.js:1574:17)
    at /名称未設定/codepositive/out/vs/base/node/id.js:131:82
    at new Promise (<anonymous>)
    at getdevDeviceId (/名称未設定/codepositive/out/vs/base/node/id.js:131:43)
    at setupServerServices (/名称未設定/codepositive/out/vs/server/node/serverServices.js:87:37)
    at createServer (/名称未設定/codepositive/out/vs/server/node/remoteExtensionHostAgentServer.js:600:103)
    at async file:///Users/kenz/src/vscode/src/server-main.js:91:20
    at async Server.<anonymous> (file:///Users/kenz/src/vscode/src/server-main.js:179:3) {
  phase: 'loading',
  moduleId: '@vscode/deviceid',
  neededBy: [ '===anonymous4===' ]
}
Web UI available at http://localhost:8000?tkn=5a62dd10-693d-4dbd-be37-058694b0f6e0
[21:33:46] Extension host agent started.
``` 

![screenshot34](https://github.com/user-attachments/assets/c6f4ebce-d209-49c3-9784-a52514484d3c)

**Comment by kenz-gelsoft** [comment]
PoC is done, but web+ (haikuwebkit) is not stable enough yet, and yarned source tree is too large (1.7GB), there are some challanges to be usable.
